File: src/lib/linkbot.py
import _linkbot

class Accelerometer:
    def __init__(self, linkbot):
        self.__linkbot = linkbot

    def set_event_handler(callback=None, granularity=0.05):
        return self.__linkbot.set_accelerometer_callback(callback)

    def values():
        return self.__linkbot.get_accelerometer()

    def x():
        return self.__linkbot.get_accelerometer[0]

    def y():
        return self.__linkbot.get_accelerometer[1]

    def z():
        return self.__linkbot.get_accelerometer[2]

class Battery:
    def __init__(self, linkbot):
        self.__linkbot = linkbot

    def percentage(self):
        return self.__linkbot.get_battery_percentage()

    def voltage(self):
        return self.__linkbot.get_battery_voltage()

class Buttons:
    def __init__(self, linkbot):
        self.__linkbot = linkbot

    def a(self):
        raise NotImplementedError("Method not implemented.")

    def b(self):
        raise NotImplementedError("Method not implemented.")

    def pwr(self):
        raise NotImplementedError("Method not implemented.")

    def set_event_handler(callback=None):
        self.__linkbot.set_button_callback(callback)

class Buzzer:
    def __init__(self, linkbot):
        self.__linkbot = linkbot

    def frequency(self):
        raise NotImplementedError("Method not implemented.")

    def set_frequency(self, frequency):
        return self.__linkbot.set_buzzer_frequency(frequency)

class Led:
    def __init__(self, linkbot):
        self.__linkbot = linkbot

    def color():
        return self.__linkbot.get_led_color()

    def set_color(self, r, g, b):
        return self.__linkbot.set_led_color(r, g, b)

    def set_color_code(self, code):
        return self.__linkbot.set_led_color_rgb(code)

class Motor:
    class Controller:
        PID = 1
        CONST_VEL = 2
        SMOOTH = 3
        ACCEL = 4

    class State:
        COAST = 0
        HOLD = 1
        MOVING = 2
        ERROR = 4

    def __init__(self, linkbot, index):
        self.__linkbot = linkbot
        self.__index = index

    def accel(self):
        raise NotImplementedError("Method not implemented.")

    def angle(self):
        return self.__linkbot.get_joint_angle(self.__index)

    def begin_accel(self, timeout, v0=0.0, state_on_timeout=State.COAST):
        raise NotImplementedError("Method not implemented.")

    def begin_move(self, timeout=0, forward=True, state_on_timeout=State.COAST):
        raise NotImplementedError("Method not implemented.")

    def controller(self):
        raise NotImplementedError("Method not implemented.")

    def decel(self):
        raise NotImplementedError("Method not implemented.")

    def move(self, angle, relative=True, wait=True):
        fn = None
        if relative:
            if wait:
                fn = self.__linkbot.move_joint
            else:
                fn = self.__linkbot.move_joint_nb
        else:
            if wait:
                fn = self.__linkbot.move_joint_to
            else:
                fn = self.__linkbot.move_joint_to_nb

        return fn(self.__index, angle)

    def move_wait(self):
        return self.__linkbot.move_joint_wait(self.__index)

    def omega(self):
        raise NotImplementedError("Method not implemented.")

    def set_power(self, power):
        return self.__linkbot.set_motor_power(self.__index, power)

class Motors:
    def __init__(self, linkbot):
        self.__linkbot = linkbot
        self.__motors = []
        for i in range(3):
            self.__motors.append(Motor(linkbot, i))

    def __getitem__(self, key):
        return self.__motors[key]

    def angles(self):
        return self.__linkbot.get_joint_angles()

    def begin_move(self, mask=0x07, timeouts=(0,0,0), forward=(True,True,True), states_on_timeout=(0,0,0), wait=True):
        raise NotImplementedError("Method not implemented.")

    def move(self, angles, mask=0x07, relative=True, timeouts=None, states_on_timeout=None, wait=True):
        if timeouts is None:
            timeouts = [None, None, None]
        if states_on_timeout is None:
            states_on_timeout = [None, None, None]
        wait_mask = 0x00
        if wait:
            wait_mask = 0x07
        return self.__linkbot._move_generic( angles[0], angles[1], angles[2],
                                             relative, relative, relative,
                                             None, None, None,
                                             timeouts[0], timeouts[1], timeouts[2],
                                             states_on_timeout[0], states_on_timeout[1], states_on_timeout[2],
                                             mask, wait_mask)

    def reset(self):
        return self.__linkbot.reset()

    def set_event_handler(self, callback, granularity=2.0):
        return self.__linkbot.set_encoder_callback(callback, granularity)

    def set_powers(powers, mask=0x07):
        return self.__linkbot.set_motor_powers(powers[0], powers[1], powers[2])

class Linkbot(_linkbot._Linkbot):
    def __init__(self, serial_id):
        _linkbot._Linkbot.__init__(self, serial_id.upper())  

        self.accelerometer = Accelerometer(self)
        self.battery = Battery(self)
        self.buttons = Buttons(self)
        self.buzzer = Buzzer(self)
        self.led = Led(self)
        self.motors = Motors(self)

    def set_joint_states(self, dir1, dir2, dir3, mask=0x07):
        return self._set_joint_states(dir1, dir2, dir3, mask)

    def move_smooth(self, angle1, angle2, angle3, mask=0x07):
        return self._move_smooth(angle1, angle2, angle3, mask)

    def move_smooth_nb(self, angle1, angle2, angle3, mask=0x07):
        return self._move_smooth_nb(angle1, angle2, angle3, mask)

    def move_to_smooth(self, angle1, angle2, angle3, mask=0x07):
        return self._move_to_smooth(angle1, angle2, angle3, mask)

    def move_to_smooth_nb(self, angle1, angle2, angle3, mask=0x07):
        return self._move_to_smooth_nb(angle1, angle2, angle3, mask)

    def set_joint_accels(self, a1, a2, a3, mask=0x07):
        return self._set_joint_accels(a1, a2, a3, mask)

    def set_joint_decels(self, a1, a2, a3, mask=0x07):
        return self._set_joint_decels(a1, a2, a3, mask)

File: src/lib/_linkbot.js
var $builtinmodule = function (name) {
"use strict";

    var mod = {};

    function __Linkbot() {
    }

    (function(proto) {
        // Getters

        proto.$get_accelerometer = function() {
            return this.inner.getAccelerometer();
        };

        proto.$get_buttons = function() {
            return this.inner.getButtons();
        };

        proto.$get_joint_angle = function(joint) {
            return this.inner.getJointAngle(joint);
        };

        proto.$get_joint_angles = function() {
            return this.inner.getJointAngles();
        };

        // Movement
       
        proto.$drive_angle = function(angle) {
            return this.inner.driveAngle(angle);
        };
      
        proto.$drive_distance = function(distance, radius) {
            return this.inner.driveDistance(distance, radius);
        };

        proto.$drive_distance_nb = function(distance, radius) {
            return this.inner.driveDistanceNB(distance, radius);
        };
        
        proto.$move = function(angle1, angle2, angle3) {
            return this.inner.move(angle1, angle2, angle3);
        };

        proto.$move_nb = function(angle1, angle2, angle3) {
            return this.inner.moveNB(angle1, angle2, angle3);
        };

        proto.$move_to = function(angle1, angle2, angle3) {
            return this.inner.moveTo(angle1, angle2, angle3);
        };

        proto.$move_to_nb = function(angle1, angle2, angle3) {
            return this.inner.moveToNB(angle1, angle2, angle3);
        };

        proto.$move_joint = function(joint, angle) {
            return this.inner.moveJoint(joint, angle);
        };

        proto.$move_joint_nb = function(joint, angle) {
            return this.inner.moveJointNB(joint, angle);
        };

        proto.$move_joint_to = function(joint, angle) {
            return this.inner.moveJointTo(joint, angle);
        };

        proto.$move_joint_to_nb = function(joint, angle) {
            return this.inner.moveJointToNB(joint, angle);
        };

        proto.$move_joint_wait = function(joint) {
            return this.inner.moveWait(1<<joint);
        };

        proto.$reset = function() {
            return this.inner.reset();
        };

        proto.$reset_to_zero = function() {
            return this.inner.resetToZero();
        };

        proto.$reset_to_zero_nb = function() {
            return this.inner.resetToZeroNB();
        };

        proto.$stop = function() {
            return this.inner.stop();
        };

        proto.$turn_angle = function(angle, radius, trackLength) {
            return this.inner.turnAngle(angle, radius, trackLength);
        };

        proto.$turn_angle_nb = function(angle, radius, trackLength) {
            return this.inner.turnAngleNB(angle, radius, trackLength);
        };

        proto.$move_wait = function() {
            return this.inner.moveWait();
        };

        // Setters

        proto.$set_buzzer_frequency = function(frequency) {
            return this.inner.setBuzzerFrequency(frequency);
        };

        proto.$set_joint_speed = function(joint, speed) {
            return this.inner.setMotorSpeeds(speed, speed, speed, 1<<joint);
        };

        proto.$set_joint_speeds = function(speed1, speed2, speed3) {
            return this.inner.setMotorSpeeds(speed1, speed2, speed3);
        };

        proto.$_set_joint_states = function(s1, s2, s3, mask) {
            return this.inner.setJointStates(s1, s2, s3, mask);
        };

        proto.$set_led_color = function(r, g, b) {
            return this.inner.setLedColor(r, g, b);
        };

        proto.$set_led_color_rgb = function(color_code) {
            return this.inner.setLedColorRgb(color_code);
        };

        proto.$set_motor_power = function(joint, power) {
            return this.inner.setMotorPowers(power, power, power, 1<<joint);
        };

        proto.$set_motor_powers = function(power1, power2, power3) {
            return this.inner.setMotorPowers(power1, power2, power3);
        };

        // Callbacks
        
        proto.$set_accelerometer_callback = function(callback) {
            if ( callback ) {
                return this.inner.setAccelerometerHandler( function( timestamp, x, y, z ) {
                    callback(timestamp, x, y, z);
                });
            } else {
                return this.inner.setAccelerometerHandler();
            }
        };

        proto.$set_button_callback = function(callback) {
            if ( callback ) {
                return this.inner.setButtonHandler( function(timestamp, buttonNo, buttonState) {
                    callback(timestamp, buttonNo, buttonState);
                });
            } else {
                return this.inner.setButtonHandler();
            }
        };

        proto.$set_encoder_callback = function(callback, granularity) {
            if ( callback ) {
                return this.inner.setEncoderHandler( function(jointNo, angle, timestamp) {
                    callback(timestamp, jointNo, angle*180.0/3.14159);
                });
            } else {
                return this.inner.setEncoderHandler();
            }
        };

    })(__Linkbot.prototype);

    // InstantPromise is a workaround to allow usage of the clean promise-style
    // then/catch syntax but to instantly call resolve the then/catch chain so we
    // can avoid creating Suspensions in unnecessary cases.  This is desirable
    // because Suspensions have a fairly large negative impact on overall
    // performance.  These 'instant promises' come into play when a tracer()
    // call is made with a value other than 1.  When tracer is 0 or greater than 1
    // , we can bypass the creation of a Suspension and proceed to the next line of
    // code immediately if the current line is not going to incur involve a screen
    // update. We determine if a real promise or InstantPromise is necessary by
    // checking FrameManager.willRenderNext()
    function InstantPromise() {
        this.lastResult = undefined;
        this.lastError  = undefined;
    }

    InstantPromise.prototype.then = function(cb) {
        if (this.lastError) {
            return this;
        }

        try {
         this.lastResult = cb(this.lastResult);
        } catch(e) {
            this.lastResult = undefined;
            this.lastError  = e;
        }

        return this.lastResult instanceof Promise ? this.lastResult : this;
    };

    InstantPromise.prototype.catch = function(cb) {
        if (this.lastError) {
            try {
                this.lastResult = cb(this.lastError);
                this.lastError  = undefined;
            } catch(e) {
                this.lastResult = undefined;
                this.lastError = e;
            }
        }

        return this.lastResult instanceof Promise ? this.lastResult : this;
    };

    function pythonToJavascriptFunction(pyValue, scope) {
        return function() {
            var argsJs = Array.prototype.slice.call(arguments),
                argsPy = argsJs.map(
                    function(argJs) {return Sk.ffi.remapToPy(argJs);}
                );

            if (typeof(scope) !== "undefined") {
                argsPy.unshift(scope);
            }
            return Sk.misceval.applyAsync(
                undefined, pyValue, undefined, undefined, undefined, argsPy
            ).catch(Sk.uncaughtException);
        };
    }

    function addModuleMethod(klass, module, method, scopeGenerator) {
        var publicMethodName = method.replace(/^\$/, ""),
            displayName      = publicMethodName.replace(/_\$[a-z]+\$$/i, ""),
            maxArgs          = klass.prototype[method].length,
            minArgs          = klass.prototype[method].minArgs,
            keywordArgs      = klass.prototype[method].keywordArgs,
            returnType       = klass.prototype[method].returnType,
            isSk             = klass.prototype[method].isSk,
            wrapperFn;

        if (minArgs === undefined) {
            minArgs = maxArgs;
        }

        wrapperFn = function() {
            var args     = Array.prototype.slice.call(arguments, 0),
                instance = scopeGenerator ? scopeGenerator() : args.shift().instance,
                i, result, susp, resolution, lengthError;

            if (args.length < minArgs || args.length > maxArgs) {
                lengthError = minArgs === maxArgs ?
                    "exactly " + maxArgs :
                    "between " + minArgs + " and " + maxArgs;

                throw new Sk.builtin.TypeError(displayName + "() takes " + lengthError + " positional argument(s) (" + args.length + " given)");
            }

            for (i = args.length; --i >= 0;) {
                if (args[i] !== undefined) {
                    if (args[i] instanceof Sk.builtin.func) {
                        args[i] = pythonToJavascriptFunction(args[i]);
                    }
                    else if (args[i] instanceof Sk.builtin.method) {
                        args[i] = pythonToJavascriptFunction(args[i].im_func, args[i].im_self);
                    }
                    else if (args[i] && args[i].$d instanceof Sk.builtin.dict && args[i].instance) {
                        args[i] = args[i].instance;
                    }
                    else {
                        args[i] = Sk.ffi.remapToJs(args[i]);
                    }
                }
            }

            try {
                result = instance[method].apply(instance, args);
            } catch(e) {
                if (window && window.console) {
                    window.console.log("wrapped method failed");
                    window.console.log(e.stack);
                }
                throw e;
            }

            if (result instanceof InstantPromise) {
                result = result.lastResult;
            }

            if (
                    (result instanceof Promise) ||
                    result.isPromise
               ) {
                result = result.catch(function(e) {
                    if (window && window.console) {
                        window.console.log("promise failed");
                        window.console.log(e.stack);
                    }
                    throw e;
                });

                susp = new Sk.misceval.Suspension();

                susp.resume = function() {
                    return (resolution === undefined) ?
                        Sk.builtin.none.none$ :
                        Sk.ffi.remapToPy(resolution);
                };

                susp.data = {
                    type: "Sk.promise",
                    promise: result.then(function(value) {
                        resolution = value;
                        return value;
                    })
                };

                return susp;
            }
            else {
                if (result === undefined) return Sk.builtin.none.none$;
                if (isSk) return result;
                if (typeof returnType === "function") {
                    return returnType(result);
                }

                return Sk.ffi.remapToPy(result);
            }
        };

        if (keywordArgs) {
            wrapperFn.co_varnames = keywordArgs.slice();
            // make room for required arguments
            for(var i = 0; i < minArgs; i++) {
                wrapperFn.co_varnames.unshift("");
            }
            if (!scopeGenerator) {
                // make room for the "self" argument
                wrapperFn.co_varnames.unshift("");
            }
        }

        module[publicMethodName] = new Sk.builtin.func(wrapperFn);
    }

    function LinkbotWrapper($gbl, $loc) {
        $loc.__init__ = new Sk.builtin.func(function (self, serial_id) {
            self.instance = new __Linkbot();
            Sk.builtin.pyCheckArgs("__init__", arguments, 2, 2);
            serial_id = Sk.ffi.remapToJs(serial_id);
            var robot_promise = getDaemon().then(function(daemon) {
                return daemon.getRobot(serial_id);
            });
            var susp = new Sk.misceval.Suspension();
            susp.resume = function() {
                if ( susp.data['error'] ) {
                    throw new Sk.builtin.RuntimeError('Could not connect to robot: ' + serial_id);
                } else {
                    return Sk.builtin.none.none$;
                }
            };
            susp.data = {
                type: "Sk.promise",
                promise: robot_promise.then(function(robot) {
                    self.instance.inner = robot;
                    return robot;
                })
            };
            return susp;
        });

        for(var key in __Linkbot.prototype) {
            if (/^\$[a-z_]+/.test(key)) {
                addModuleMethod(__Linkbot, $loc, key);
            }
        }
    };

    mod._Linkbot = Sk.misceval.buildClass(mod, LinkbotWrapper, "_Linkbot", []);

    return mod;
};
